### PR TITLE
(BKR-988) Add Test Tagging OR Logic

### PR DESCRIPTION
**Pull request by kevpl**
This PR refactors test tagging logic into its own module, & adds the OR CLI parameter.

This PR will probably need to be squished, but I wanted to have it separated out at first, so the refactor could be looked at separately from the additional functionality.

### CODE REVIEW

**Comment by kevpl** [comment]
@cthorn42 I know you were dealing with test tagging a while ago. If you're still dealing with that, please take a look at these changes that could be coming your way. If not, would you point out who is working with that now, so that we can make sure they're aware of these changes?

**Comment by cthorn42** [comment]
@kevpl I'll take a look at this, specifically test-tag-or is a lot cleaner then what I'm currently using for the nightly pe_acceptance_tests.

**Comment by tvpartytonight** [comment]
@puppetlabs-jenkins retest this please

**Comment by kevpl** [comment]
@cthorn42 any news on your review?

File: lib/beaker/options/command_line_parser.rb [hunk 1]
```diff
@@ -245,12 +245,27 @@ def initialize
...
+                  'DEPRECATED - use --test-tag-and instead' do |value|
+            @cmd_options[:test_tag_and] = value
+          end
+          opts.on '--test-tag-and TAGS',
+                  'Run the set of tests matching ALL of the provided single or comma separated list of tags' do |value|
+            @cmd_options[:test_tag_and] = value
+          end
+
+          opts.on '--test-tag-or TAGS',
+                  'Run the set of tests matching ANY of the provided single or command separated list of tags' do |value|
```
Comment (kevpl) [inline lib/beaker/options/command_line_parser.rb]: command

File: lib/beaker/options/command_line_parser.rb [hunk 2]
```diff
@@ -245,12 +245,27 @@ def initialize
...
           end
 
-          opts.on '--exclude-tag TAGS', 'Run the set of tests that do not contain ANY of the provided single or command separated list of tags' do |value|
-            @cmd_options[:tag_excludes] = value
+          opts.on '--exclude-tag TAGS',
+                  'DEPRECATED - use --test-tag-exclude instead' do |value|
+            @cmd_options[:test_tag_exclude] = value
+          end
+          opts.on '--test-tag-exclude TAGS',
+                  'Run the set of tests that do not contain ANY of the provided single or command separated list of tags' do |value|
```
Comment (kevpl) [inline lib/beaker/options/command_line_parser.rb]: command

File: lib/beaker/options/parser.rb [hunk 3]
```diff
@@ -411,13 +415,15 @@ def check_hypervisor_config(visor)
...
-        @options[:tag_includes] ||= ''
-        @options[:tag_excludes] ||= ''
-        @options[:tag_includes] = @options[:tag_includes].split(',') if @options[:tag_includes].respond_to?(:split)
-        @options[:tag_excludes] = @options[:tag_excludes].split(',') if @options[:tag_excludes].respond_to?(:split)
-        @options[:tag_includes].map!(&:downcase)
-        @options[:tag_excludes].map!(&:downcase)
+      # @note refer to {Beaker::DSL::TestTagging} for test tagging implementation
+      #
+      def normalize_test_tags!
+        @options[:test_tag_and]     ||= ''
```
Comment (kevpl) [inline lib/beaker/options/parser.rb]: or

File: lib/beaker/options/presets.rb [hunk 4]
```diff
@@ -34,8 +34,8 @@ class Presets
         :release_yum_repo_url => ['BEAKER_RELEASE_YUM_REPO', 'RELEASE_YUM_REPO'],
         :dev_builds_url       => ['BEAKER_DEV_BUILDS_URL', 'DEV_BUILDS_URL'],
         :vbguest_plugin       => ['BEAKER_VB_GUEST_PLUGIN', 'BEAKER_vb_guest_plugin'],
-        :tag_includes         => ['BEAKER_TAG'],
-        :tag_excludes         => ['BEAKER_EXCLUDE_TAG'],
+        :test_tag_and         => ['BEAKER_TAG', 'BEAKER_TEST_TAG_AND'],
+        :test_tag_exclude     => ['BEAKER_EXCLUDE_TAG', 'BEAKER_TEST_TAG_EXCLUDE'],
```
Comment (kevpl) [inline lib/beaker/options/presets.rb]: or

File: lib/beaker/options/parser.rb (hunk 3 above)
Comment (kevpl) [inline lib/beaker/options/parser.rb]: added below

**Comment by tvpartytonight** [comment]
This looks good to me! 👍

**Pull request merged**